Review fixes for "Make play() run in a separate thread (#1500)"

MichaelOrlov opened the pull request:
- Relates https://github.com/ros2/rosbag2/pull/1503
- Addressed multiple race conditions in player and some renames after review
- Fixed multiple race conditions in tests when wait_for_playback_to_end() calls in a separate thread before
Player::play() method. Those race conditions would likely lead to false negative or flaky tests.
- Replaced player_future with wait_for_playback_to_end() to the direct call of the player_->wait_for_playback_to_end() in the most of the places.
- Renamed `wait_for_playback_to_end()` to the `wait_for_playback_to_finish()`
- Use rcpputils::unique_lock for RCPPUTILS TSA guarded mutexes
   - Rationale: To support Clang Thread Safety Analysis
- Removed locally defined TSAUniqueLock which is a duplicate of the rcpputils::unique_lock.
- Removed some unused includes in the `player.cpp`
- Add timeout parameter to the wait_for_playback_to_finish()
   - Rationale: Make wait_for_playback_to_finish(..) to be optionally as a non-blocking call. By default, it could be called without a parameter and will be a blocking call as before.

morlov-apexai committed:
Address multiple race conditions in player and some renames after review

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Fixes in the rosbag2_transport tests

- Fixed multiple race conditions in tests when
wait_for_playback_to_end() calls in a separate thread before
Player::play() method. Those race conditions would likely lead to a
false negative or flaky tests.

- Replaced player_future with wait_for_playback_to_end() to the direct
call of the player_->wait_for_playback_to_end() in the most of the
places.

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
cc: @roncapat still WIP.

roncapat commented:
First of all, as always, thank you for the review :)

I have one doubt: wouldn't calling `test_fixture_->player_->wait_for_playback_to_end();` after `play()` make possible the case in which the player thread finishes _before_ the `wait_for_playback_to_end()`call, and so the condition variable is being notified before waiting on it?

MichaelOrlov commented:
Yes, it could be that the player thread finishes before the `wait_for_playback_to_end()` call. But this is a valid case and it is properly handled in the `wait_for_playback_to_end()`.
The
```
playback_finished_cv_.wait(is_in_playback_lk, [this] {return !is_in_playback_.load();});
```
will return immediately without waiting if atomic `is_in_playback_` equals false at the moment of its calling.
There are no race conditions since now `is_in_playback_` enforced to be protected by the mutex and we will have `RCPPUTILS_TSA_GUARDED_BY` attribute for that.

roncapat commented:
@MichaelOrlov all clear, I was missing that bit.

morlov-apexai committed:
Rename wait_for_playback_to_end() to the wait_for_playback_to_finish()

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Use rcpputils::unique_lock for RCPPUTILS TSA guarded mutexes

- Rationale: To support Clang Thread Safety Analysis
- Removed locally defined TSAUniqueLock which is a duplicate of the
rcpputils::unique_lock.
- Also removed some unused includes in player.cpp and player.hpp

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

morlov-apexai committed:
Add timeout parameter to the wait_for_playback_to_finish(timeout)

- Rationale: Make wait_for_playback_to_finish(..) to be optionally as
a non-blocking call

Signed-off-by: Michael Orlov <michael.orlov@apex.ai>

MichaelOrlov commented:
@roncapat This PR is ready for review and perhaps for the merge if no concerns.
